### PR TITLE
Add Gemma 4 Fixes

### PR DESCRIPTION
**Pull request by N8python**
Standard codex-generated warning applies here - this was done by having codex verify KL against pytorch ref and having it hunt down implementation bugs. I take full responsibility for all code generated.

List of changes:
- gemma4.py (line 55): Fixed to distinguish language_model.layers from already-converted language_model.model.layers - this makes directly converting from `google/gemma-4-31B-it` viable.
- changes gemma4_text to use a new ProportionalRope class w/ proper head-rotation (instead of rotating prefix of head, rotate r/2 dims of left half, r/2 dims of right half)
- adds additional support for input_embeddings so token ids can be recovered (does this by exact matching input_embeddings, inefficient but this is only for the path where you call an E2B/E4B model with input_embeddings only - this is an extreme edge case). 
- adds tests for all these.

### CODE REVIEW

File: mlx_lm/models/gemma4.py [hunk 1]
```diff
@@ -66,10 +70,13 @@ def sanitize(self, weights):
...
-            # HF language_model maps to Gemma4TextModel directly (no lm_head wrapper),
-            # but our Model wraps it in self.model, so prepend "model." to each key
+            # Converted MLX weights already use "model.layers.*" while raw HF
+            # checkpoints use "layers.*" under language_model.
             lm_weights = dict(tree_flatten(new_weights["language_model"]))
-            lm_weights = {"model." + k: v for k, v in lm_weights.items()}
+            if not any(
+                k.startswith("model.") or k.startswith("lm_head.") for k in lm_weights
+            ):
+                lm_weights = {"model." + k: v for k, v in lm_weights.items()}
```
Comment (Blaizzy) [inline mlx_lm/models/gemma4.py:79]: this is fine, just make sure to test with quants and full precision

**Review by Blaizzy — APPROVED** [review]
LGTM, thanks!
